docs/HARDWARE-BRINGUP.md

```diff
@@ -50,6 +50,108 @@ One word of caution: a lot of ARM board (like the ever popular Raspberry Pi) act
+
+The Linux kernel view of the device hierarchy created based on the information presented in a device tree is available under `/sys/devices/platform`.
+
+### 4.2. Bus discovery
+
+Having a manifest of all the hardware present on the system is only as useful as CPU's ability to talk to all that hardware. Some of the hardware can be driven directly by memory mapped registers and DMA, other require buses to be available. A bus itself is a kind of a hardware device (at least the end of the bus that connects it to the host CPU) and Linux kernel tracks all the buses it knows about under `/sys/bus/`. Before you can move to enabling a driver, make sure that the required bus is visible under `/sus/bus/BUS-NAME`. The most interesting subfolder under that is `/sus/bus/BUS-NAME/devices`. This is Linux kernel's idea of what devices sit on that bus. Sophisticated buses (like USB) allow for dynamic device enumeration and discovery so that `/sus/bus/usb/devices` is maintained up-to-date by the bus driver itself. Simpler buses (like I2C) don't allow for device discovery. The device sitting on such a bus has to either be described in an ACPI table or device tree entry OR it has to be [manually instantiated](https://www.kernel.org/doc/html/latest/i2c/instantiating-devices.html). For example, the following will instantiate device with the name `eeprom` at I2C address `0x50` on an I2C bus #3:
+
+```console
+echo eeprom 0x50 > /sys/bus/i2c/devices/i2c-3/new_device
+```
+
+When dealing with any bus (even sophisticated ones like USB) tools that aid in probing and discovery are a must (you will find most of them in EVE's debug container):
+
+* [lspci](https://pkgs.alpinelinux.org/package/edge/main/x86_64/pciutils) for PCI discovery
+* [lsusb](https://pkgs.alpinelinux.org/package/edge/main/x86/usbutils) for USB discovery
+* [i2c-tools](https://pkgs.alpinelinux.org/package/edge/community/x86_64/i2c-tools) for I2C discovery
+* [spi-tools](https://pkgs.alpinelinux.org/package/edge/community/x86_64/spi-tools) for SPI discovery
+
+If all else fails you may need to write [your own client](https://www.kernel.org/doc/Documentation/i2c/writing-clients).
+
+### 4.3 Pin control
+
+Before PCIe, memory mapped registers and DMA - hardware could only be accessed by a CPU via pins on its package also know as GPIO. Electrically, a given pin can be either at 0 or 1 voltage level and CPU has to ways of setting these values:
```

Comment:
"to ways" probably should be "two ways".

Reply:
fixed